partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelPeeking.kt

```diff
@@ -2,44 +2,20 @@ package org.partiql.eval.internal.operator.rel
 
 import org.partiql.eval.internal.Environment
 import org.partiql.eval.internal.Record
+import org.partiql.eval.internal.helpers.IteratorPeeking
 import org.partiql.eval.internal.operator.Operator
 
 /**
  * For [Operator.Relation]'s that MUST materialize data in order to execute [hasNext], this abstract class caches the
  * result of [peek] to implement both [hasNext] and [next].
  */
-internal abstract class RelPeeking : Operator.Relation {
-
-    private var _next: Record? = null
-
-    /**
-     * @return Null when there is not another record to be produced. Returns a [Record] when able to.
-     *
-     * @see RelPeeking
-     */
-    abstract fun peek(): Record?
+internal abstract class RelPeeking : Operator.Relation, IteratorPeeking<Record>() {
 
     override fun open(env: Environment) {
-        _next = null
-    }
-
-    override fun hasNext(): Boolean {
-        if (_next != null) {
-            return true
-        }
-        this._next = peek()
-        return this._next != null
-    }
-
-    override fun next(): Record {
-        val next = _next
-            ?: peek()
-            ?: error("There was not a record to be produced, however, next() was called. Please use hasNext() beforehand.")
-        this._next = null
-        return next
+        next = null
     }
 
     override fun close() {
-        _next = null
+        next = null
     }
```

Comment:
nit, but it would be better to have the abstract method hold the close control so that inheritors don't need to call super.close(). In some languages you can do something like [@mustCallSuper](https://github.com/vandadnp/flutter-tips-and-tricks/blob/main/tipsandtricks/mustcallsuper-in-dart/mustcallsuper-in-dart.md) but we can do better with a concrete close that does

```
override fun close() {
     close1()
     next = null
}

abstract fun close1()
```

It's a nit, but I wouldn't want to track down a bug because an inheritor didn't call super. Perhaps we just add a comment to keep the change small.

Reply:
Agreed. I was going to do that in this PR, but I didn't want the diff to affect implementers of RelPeeking. Seeing that you've reviewed the rest of the PR, I'll update. Thanks!